chips/rp2040/src/usb.rs

```diff
@@ -0,0 +1,2532 @@
+            self.registers.usb_muxing.set(0);
+            self.registers
+                .usb_muxing
+                .modify(USB_MUXING::TO_DIGITAL_PAD::SET + USB_MUXING::SOFTCON::SET);
+            for _i in 0..106400 {
+                cortexm0p::support::nop()
+            }
+            self.registers.usb_muxing.set(0);
+            self.registers
+                .usb_muxing
+                .modify(USB_MUXING::TO_PHY::SET + USB_MUXING::SOFTCON::SET);
+            p.finish_usb_errata(prev_ctrl, prev_pad);
+        });
+    }
+
+    fn handle_bus_reset(&self) {
+        for (ep, desc) in self.descriptors.iter().enumerate() {
+            match desc.state.get() {
+                EndpointState::Disabled => {}
+                EndpointState::Ctrl(_) => desc.state.set(EndpointState::Ctrl(CtrlState::Init)),
+                EndpointState::Bulk(transfer_type, in_state, out_state) => {
+                    desc.state.set(EndpointState::Bulk(
+                        transfer_type,
+                        in_state.map(|_| BulkInState::Init),
+                        out_state.map(|_| BulkOutState::Init),
+                    ));
+                    if out_state.is_some() {
+                        self.dpsram.ep_buf_ctrl[ep].ep_out_buf_ctrl.set(0);
+                        self.dpsram.ep_buf_ctrl[ep].ep_out_buf_ctrl.modify(
+                            EP_BUFFER_CONTROL::AVAILABLE0::SET
+                                + EP_BUFFER_CONTROL::TRANSFER_LENGTH0.val(64 as u32),
+                        );
+                    }
+                }
+            }
+            self.next_pid_in[ep].set(0);
+            self.next_pid_out[ep].set(0);
+            desc.request_transmit_in.set(false);
+            desc.request_transmit_out.set(false);
+        }
+        self.dpsram.ep_buf_ctrl[0]
+            .ep_out_buf_ctrl
+            .modify(EP_BUFFER_CONTROL::DATA_PID0::SET);
+
+        self.dpsram.ep_buf_ctrl[0].ep_in_buf_ctrl.modify(
+            EP_BUFFER_CONTROL::AVAILABLE0::SET
+                + EP_BUFFER_CONTROL::TRANSFER_LENGTH0.val(64)
+                + EP_BUFFER_CONTROL::DATA_PID0::CLEAR,
+        );
+        self.registers.buff_status.set(0);
+        self.registers.addr_endp.modify(ADDR_ENDP::ADDRESS.val(0));
+
+        self.address.set(0);
+        self.client.map(|client| {
+            client.bus_reset();
+        });
+        self.registers.sie_status.modify(SIE_STATUS::BUS_RESET::SET);
+    }
+
+    pub fn handle_interrupt(&self) {
+        if self.registers.ints.is_set(INTS::BUS_RESET) {
+            self.handle_bus_reset();
+        }
+
+        if self.registers.buff_status.get() != 0 {
+            self.handle_buff_status();
+        }
+
+        self.process_requests();
+
+        if self.registers.ints.is_set(INTS::SETUP_REQ) {
+            self.registers.sie_status.modify(SIE_STATUS::SETUP_REC::SET);
+            self.usb_handle_setup_packet();
+        }
+    }
+
+    fn handle_buff_status(&self) {
+        for i in 0..31 {
```

Comment:
Out of curiosity, why do we need the loop here? looks like all the match branches run?

Reply:
Your are right, the match has no meaning here.

Reply:
It works.